### PR TITLE
Fixes #13667: Add condition_once and execute_once generic methods

### PR DESCRIPTION
**Pull request by amousset**
https://www.rudder-project.org/redmine/issues/13667

### CODE REVIEW

**Comment by amousset** [comment]
Do not merge for now, wip

**Comment by amousset** [comment]
Commit modified

**Comment by amousset** [comment]
Commit modified

**Comment by amousset** [comment]
Commit modified

**Comment by amousset** [comment]
Commit modified

File: tree/30_generic_methods/command_execution_once.cf [hunk 1]
```diff
@@ -0,0 +1,96 @@
...
+    pass3::
+      "clean"   usebundle => _classes_sanitize("${old_class_prefix}");
+      "clean"   usebundle => _classes_sanitize("${class_prefix}");
+
+      "report"
+        usebundle  => _log_v3("Execute the command '${command}' once", "${command}", "${old_class_prefix}", "${class_prefix}", @{args});
+
+  commands:
+    reserved::
+      "${command}"
```
Comment (ncharles) [inline tree/30_generic_methods/command_execution_once.cf]: why not using command_execution ?

Comment (amousset) [reply tree/30_generic_methods/command_execution_once.cf]: Makes sense indeed (with `command_execution_result`)

**Comment by amousset** [comment]
Commit modified

**Comment by amousset** [comment]
Commit modified

**Comment by amousset** [comment]
Commit modified

**Comment by amousset** [comment]
Commit modified

**Comment by amousset** [comment]
Commit modified

**Comment by amousset** [comment]
Commit modified

File: tests/acceptance/30_generic_methods/command_execution_once.cf [hunk 2]
```diff
@@ -0,0 +1,85 @@
...
+
+    "phC"  usebundle => command_execution_once("exit 2",  "${ok_codes}", "ok", "testC");
+    "phCC" usebundle => command_execution_once("exit 2 ", "${ok_codes}", "ok", "testC");
+
+    "phD"    usebundle => command_execution_once("exit 3",  "${ok_codes}", "ok", "testD");
+    "phDD"   usebundle => command_execution_once("exit 3 ", "${ok_codes}", "ok", "testD");
+    "phDDD"  usebundle => command_execution_once("exit 4",  "${ok_codes}", "ok", "testD");
+
+    "enable"  usebundle => set_dry_run_mode("true");
+    "phE"     usebundle => command_execution_once("exit 5",  "${ok_codes}", "ok", "testE");
```
Comment (ncharles) [inline tests/acceptance/30_generic_methods/command_execution_once.cf]: auditing execute_once is an error ? this is surprising, but i don't know what it should be

Comment (amousset) [reply tests/acceptance/30_generic_methods/command_execution_once.cf]: This is the current behavior with other command methods, I think it's the best we can do.

Comment (amousset) [reply tests/acceptance/30_generic_methods/command_execution_once.cf]: I'm adding a test for already executed command in audit mode

**Comment by amousset** [comment]
Commit modified

File: tests/acceptance/30_generic_methods/command_execution_once.cf (hunk 2 above)
Comment (amousset) [reply tests/acceptance/30_generic_methods/command_execution_once.cf]: done

**Comment by amousset** [comment]
Commit modified

**Comment by amousset** [comment]
Commit modified

**Comment by amousset** [comment]
Commit modified

File: tree/30_generic_methods/permissions_recurse.cf [hunk 3]
```diff
@@ -54,7 +54,7 @@ bundle agent permissions_recurse(path, mode, owner, group)
       "reenable_reporting_${class_prefix}"
                                    usebundle => enable_reporting,
                                   ifvarclass => "should_report";
-      "new result classes"         usebundle => _classes_copy("${inner_class}", "${class_prefix}");
+      "new result classes"         usebundle => _classes_copy("${inner_class_prefix}", "${class_prefix}");
```
Comment (ncharles) [inline tree/30_generic_methods/permissions_recurse.cf]: this should not be part of this PR

File: tree/20_cfe_basics/flag_lib.cf [hunk 4]
```diff
@@ -169,16 +198,17 @@ bundle agent set_flag(name) {
       "notreserved"
         bundle_return_value_index => "1",
                           comment => "${flag_unique.value}",
-                               if => "!flag_reserved_${cname}";
+                               if => "!reserved";
```
Comment (ncharles) [inline tree/20_cfe_basics/flag_lib.cf]: notreserved is never used - is it for future usage ?

File: tree/30_generic_methods/command_execution_once.cf [hunk 5]
```diff
@@ -0,0 +1,136 @@
...
+# Will retry to run `command -a -t` until it returns `0`. Then it will
+# not execute it again.
+#
+# @parameter command        The command to run
+# @parameter ok_codes       List of codes that produce a repaired status separated with commas (ex: 1,2,5). Defaults to 0.
+# @parameter_constraint ok_codes "allow_empty_string" : true
+# @parameter until          Try to execute the command until a particular state: 'ok', 'any' (defaults to 'any')
+# @parameter_constraint until "allow_empty_string" : true
+# @parameter_constraint until "select" : [ "", "any", "ok" ]
+# @parameter unique_id      To identify the command (allows to change the command without losing track)
```
Comment (ncharles) [inline tree/30_generic_methods/command_execution_once.cf]: shouldn't it be optionnal, and default to command ?

Comment (amousset) [reply tree/30_generic_methods/command_execution_once.cf]: would be better indeed!

File: tree/20_cfe_basics/flag_lib.cf [hunk 6]
```diff
@@ -49,6 +49,8 @@
 #   * `ok`: if the flag was correctly reserved
 #   * `failed`: if the flag could not be reserved. I can happen if it was reserved by another agent.
 #
+# * `cancel_flag` to remove a flag reservation. This does not affect already set classes.
```
[the remark this answers is not in the document]
Comment (amousset) [reply tree/20_cfe_basics/flag_lib.cf]: YES, good spot, s/classes/flags

File: tree/30_generic_methods/permissions_recurse.cf (hunk 3 above)
Comment (amousset) [reply tree/30_generic_methods/permissions_recurse.cf]: definitely, I'll open a ticket (I just saw this while copy/pasting the inner class code)

**Comment by amousset** [comment]
Commit modified

File: tree/20_cfe_basics/flag_lib.cf (hunk 4 above)
Comment (amousset) [reply tree/20_cfe_basics/flag_lib.cf]: It is actually used for more precise testing (and maybe for future usage too).

**Comment by amousset** [comment]
Commit modified

File: tree/20_cfe_basics/flag_lib.cf (hunk 4 above)
Comment (ncharles) [reply tree/20_cfe_basics/flag_lib.cf]: Great ! Thank you
